mosqagent.h

```diff
@@ -14,6 +14,91 @@
 struct mosqagent_idle_list;
 struct mosqagent_config;
 
+
+/**
+ * \brief De-allocator function pointer
+ *
+ * The de-allocator is responsible for cleaning up a pointer and possible
+ * heap structures in the pointed-to struct. The exact handling is based
+ * on the struct, so the user (or the library) has to provide the correct
+ * de-allocator for each memory object.
+ *
+ * Providing `NULL` as de-allocator reference will be interpreted as
+ * "no de-allocation", i.e. the object will be left alone. Use this for stack
+ * pointers of if clean-up is handled somewhere else.
+ *
+ * \note The de-allocator is also responsible for cleaning up the pointed-to
+ * memory. The pendant for `malloc`-created objects would be a pointer to
+ * `free`.
+ */
+typedef void (mqtta_mo_deallocator)(void*);
+
+/**
+ * \brief A memory object that stores ownership information.
+ *
+ * Never manipulate the values directly! Always use `mqtta_mo_move` and
+ * `mqtta_mo_set` to change pointers. Ownership information is managed
+ * automatically by these functions.
+ */
+struct mqtta_memory_object {
+    void *ptr;
+    mqtta_mo_deallocator *deallocator;
+};
+
+/**
+ * \brief Get the pointer from a memory object.
+ */
+void* mqtta_mo_ptr(const struct mqtta_memory_object *mo);
+
+/**
+ * \brief Set the pointer and transfer ownership by providing a de-allocator.
+ *
+ * Ownership transfer implies that the pointer will be managed be the entity
+ * owning the memory object, i.e. the pointer should not be manipulated
+ * afterwards, especially free must not be called. The new owner will call
+ * the de-allocator when the memory object is no longer used.
+ *
+ * Validity of the pointer is tied to the implementation of the functions using
+ * the memory object.
+ *
+ * \warning Never transfer ownership for stack pointers and global variables!
+ *
+ * \note Provding a `NULL` de-allocator is equivalent to not transferring
+ *       ownership, i.e. the memory object will not be freed.
+ * \note Provide `free` as the standard de-allocator for `malloc`-created
+ *       objects.
+ */
+struct mqtta_memory_object* mqtta_mo_move(struct mqtta_memory_object *mo,
+                                          void *ptr,
+                                          mqtta_mo_deallocator *deallocator);
+
```

Comment:
You might want to add a section `\return` to explain what is returned here. Is it the same pointer you pass or something different?